aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -0,0 +1,349 @@
+    for (int i = 0; i < inputInfo.dims; ++i) {                            \
+      inputSlices *= inputInfo.sizes[i];                                  \
+    }                                                                     \
+    int64_t topKSlices = 1;                                               \
+    for (int i = 0; i < topKInfo.dims; ++i) {                             \
+      topKSlices *= topKInfo.sizes[i];                                    \
+    }                                                                     \
+                                                                          \
+    dim3 grid;                                                            \
+    if (!THC_getGridFromTiles(inputSlices, grid)) {                       \
+      THError("Slice to sort is too large");                              \
+    }                                                                     \
+                                                                          \
+    dim3 block(std::min(THCRoundUp(sliceSize, (int64_t) C10_WARP_SIZE), (int64_t) 1024)); \
+                                                                          \
+    /* This is used as a template parameter to calculate indices. */      \
+    /* We only specialize it if all collapsed dim sizes are the */        \
+    /* same; otherwise, we use -1 which is the specialization */          \
+    /* parameter for arbitrary dimensions */                              \
+    int allDims = inputInfo.dims;                                         \
+    if (topKInfo.dims != allDims || indicesInfo.dims != allDims) {        \
+      allDims = -1;                                                       \
+    }                                                                     \
+                                                                          \
+    RUN_DIM(INDEX_T);                                                     \
+  });
+
+  // the below is safe with 0-dimensional tensors because it is based on
+  // TensorInfo which implicitly expands to 1-dimensional.
+  if (input.numel() > 0) {
+    // Based on required index size, run the algorithm with the
+    // appropriate index type
+    if (at::cuda::detail::canUse32BitIndexMath(input) &&
+        at::cuda::detail::canUse32BitIndexMath(values) &&
+        at::cuda::detail::canUse32BitIndexMath(indices)) {
+      RUN_T(uint32_t);
+    } else {
+      RUN_T(uint64_t);
+    }
+  }
+#undef RUN_T
+#undef RUN_DIM
+#undef RUN_DIR
+#undef RUN_K
+
+  // Sort the results if the user wants them sorted, since our
+  // selection routine does not ensure sorting
+  if (sorted && values.numel() > 1) {
+    // FIXME: the k/v inplace sort along slice only works for size <=
+    // 2048 at the moment
+    // Workaround:
+#if defined(THC_REAL_IS_DOUBLE) || defined(THC_REAL_IS_LONG)
+    int maxSliceSize = 1024;
+#else
+    int maxSliceSize = 2048;
+#endif
```

Comment:
THC_REAL_IS_DOUBLE is not defined in ATen, it is only used in legacy code

Reply:
Is there a replacement for this or should we just assume a value for `maxSliceSize`?

Reply:
You can do something like, for example:
```C++
if (input.element_size() >= 8) {
    maxSliceSize = 1024;
} else {
   maxSliceSize = 2048;
}
```